### ISSUE TITLE
Problem when computing velocity from TLE

### ISSUE THREAD

**Issue by daniestevez** [skyfielders/python-skyfield#224]
I believe there is a problem when computing the velocity vector of a satellite using a TLE.
[This notebook](https://github.com/daniestevez/jupyter_notebooks/blob/master/eshail2/Skyfield%20TLE%20velocity%20bug.ipynb) shows some insight into the bug. The problem seems to be with the transformation from ITRF to GCRS.

**Comment by brandon-rhodes** [comment]
Thank you for the detailed notebook! It made it easy for me to get a test script up and running, and after a couple of hours of `print()` calls and puzzling over rotation matrices, I think I've arrived at a solution, which will be included in Skyfield's next release. Thanks for drawing my attention to this error!

**Comment by brandon-rhodes** [comment]
Well, drat. that broke the test `test_iss_against_horizons()` by moving Skyfield away from agreement with HORIZONS. How did that happen? I'll post further updates as events warrant.

**Comment by brandon-rhodes** [comment]
I have reverted my attempted fix c80c80a9713b059cb5645123e0482c4a221bbb0a. Had I been paying more attention, I would have noticed that while it happened to reduce the error in the _z_ coordinate, it simultaneously increased the _x_ and _y_ errors up to the same level the _z_ had been at. So it was not overall an improvement.

I am not convinced, having looked over the numbers again, that the notebook shows a real problem. The satellite moves very slowly in the ITRF reference frame — 0.0045 km/s when I run the `skyfield.functions` routine `length_of()` on it — and so the errors on that vector are correspondingly small. But when transformed to GCRS, the vector becomes quite large since the satellite is moving quickly to keep the same bit of Earth underneath it: 3.07 km/s.

I actually see a bigger proportional error in the ITRF coordinate velocities, around 1% for the _y_ velocity, which is 10 times larger than the _z_ coordinate error you've found if it's measured against the size of that much bigger velocity vector.

Here's my script. I'm going to re-close this issue for now on the assumption that the error you've uncovered is reasonable given how large the velocity vector is, but feel free to re-open if you think you see a way to correct the coordinate transform without breaking the ISS test against HORIZONS results. In any case, thanks for asking about a puzzling behavior!
```
import skyfield.api
from skyfield.sgp4lib import EarthSatellite
from skyfield.constants import AU_KM, DAY_S
from skyfield.functions import length_of

sat = EarthSatellite('1 43700U 18090A   18335.89431171 +.00000133 +00000-0 +00000-0 0  9993',\
                     '2 43700 000.0858 245.4352 0001094 006.6237 164.6135 01.00274015000309')

print(sat.epoch.utc)
ts = skyfield.api.load.timescale()
t0 = ts.utc(2018, 12, 1, 21, 0, 0)
t1 = ts.utc(2018, 12, 1, 21, 0, 1)

print('derivative of t.MT:')
print(t1.MT - t0.MT)

print('')

#print(sat)
# print(sat.positives)
# print(sat.negatives)

rv0 = sat.at(t0)
rv1 = sat.at(t1)

mag = length_of(rv0.velocity.km_per_s)
print('magnitude of velocity vector:', mag)
error = rv1.position.km - rv0.position.km - rv0.velocity.km_per_s
print('error', error)
print('error / mag', error / mag)
print(rv1.position.km)
print(rv1.position.km[2])
print(rv0.position.km[2])
# print(rv1.position.km[2] - rv0.position.km[2])
# print(rv0.velocity.km_per_s[2])
print(rv1.position.km - rv0.position.km)
print(rv0.velocity.km_per_s)

print('')

rv0_ITRF = sat.ITRF_position_velocity_error(t0)#[:2]
r0_ITRF = rv0_ITRF[0] * AU_KM
v0_ITRF = rv0_ITRF[1] * AU_KM / DAY_S
rv1_ITRF = sat.ITRF_position_velocity_error(t1)#[:2]
r1_ITRF = rv1_ITRF[0] * AU_KM
v1_ITRF = rv1_ITRF[1] * AU_KM / DAY_S

mag = length_of(v0_ITRF)
print('magnitude of velocity vector:', mag)
error = r1_ITRF - r0_ITRF - v0_ITRF
print('error', error)
print('error / mag', error / mag)
print(r1_ITRF)
print(r1_ITRF[2])
print(r0_ITRF[2])
print(r1_ITRF[2] - r0_ITRF[2])
print(v0_ITRF[2])
```

**Comment by daniestevez** [comment]
Thank you so much for having a detailed look at this.

Maybe I should add some background for why I am interested by this issue. I was trying to study the Doppler of GEO satellites and I found that when doing the computation in the obvious way (projection of velocity vector onto line of sight vector), the result had a large error. If instead the velocity vector was approximated as the difference quotient of positions, then the result was much better.

I have updated the [notebook](https://github.com/daniestevez/jupyter_notebooks/blob/master/eshail2/Skyfield%20TLE%20velocity%20bug.ipynb) to show an example of this.

You make a fair point that the error in the velocity vector in GCRS is acceptable because the relative error is small (since the velocity vector is large). However, when computing the Doppler, the velocity vector is almost perpendicular to the line of sight vector, so we end up with a huge relative error.

Maybe this is just a problem in the approach. In the case of GEO satellites, probably it is not appropriate to use GCRS coordinates to compute Doppler (which is precisely what the `at` method does). It seems more reasonable to use ITRF coordinates instead, but I'm not sure that there is a straightforward way to do this in Skyfield. What do you think?

**Comment by brandon-rhodes** [comment]
Yes, as a first guess, I'd say stick with ITRF coordinates. Now that you've updated your notebook (thanks!), I'll re-open this issue so I remember to go look — which will probably be in a couple of weeks, since I'm at Python conferences both of the coming two weekends.

**Comment by brandon-rhodes** [comment]
Thanks to an email exchange with another Skyfield user, I was spurred to dive into this again, and I believe I have resolved the problem in 3c5b869, which should be in the release of Skyfield that I make later today or this weekend. You are welcome to try it out immediately, though, with:
```
pip install https://github.com/skyfielders/python-skyfield/archive/master.zip
```
Either way, feel free to make further comments here if you don’t see any difference or the problem seems to linger. Thanks!

**Comment by daniestevez** [comment]
Thank you for revisiting this!

I have updated Skyfield and re-run the [notebook](https://github.com/daniestevez/jupyter_notebooks/blob/master/eshail2/Skyfield%20TLE%20velocity%20bug.ipynb) and it seems that the problem disappears.

It still bugs me a little that when I compute the Doppler using the velocity vector from Skyfield vs. when I compute it using a secant vector from two positions, the results have a rather large difference of 20%. (Compare cells [12] and [14]). This may be just a problem caused by approximating the tangent vector by the secant vector. And in any case, this simple way is not appropriate for computing the Doppler very precisely, since it doesn't take light-time delay and (maybe) the fact that light moves in straight lines only in ECI coordinates into account.

So I'm happy that the result seems more or less accurate.
